kube-apiserver/tls-config.md

```diff
@@ -0,0 +1,366 @@
+	// reduced.
+	//
+	// +unionDiscriminator
+	// +optional
+	Type TLSProfileType `json:"type"`
+	// old is a TLS security profile based on:
+	//
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Old_backward_compatibility
+	//
+	// +optional
+	// +nullable
+	Old *OldTLSProfile `json:"old,omitempty"`
+	// intermediate is a TLS security profile based on:
+	//
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
+	//
+	// +optional
+	// +nullable
+	Intermediate *IntermediateTLSProfile `json:"intermediate,omitempty"`
+	// modern is a TLS security profile based on:
+	//
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility
+	//
+	// +optional
+	// +nullable
+	Modern *ModernTLSProfile `json:"modern,omitempty"`
+	// custom is a user-defined TLS security profile. Be extremely careful using a custom
+	// profile as invalid configurations can be catastrophic. An example custom profile
+	// looks like this:
+	//
+	// +optional
+	// +nullable
+	Custom *CustomTLSProfile `json:"custom,omitempty"`
+}
+
+// OldTLSProfile is a TLS security profile based on:
+// https://wiki.mozilla.org/Security/Server_Side_TLS#Old_backward_compatibility
+type OldTLSProfile struct{}
+
+// IntermediateTLSProfile is a TLS security profile based on:
+// https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
+type IntermediateTLSProfile struct{}
+
+// ModernTLSProfile is a TLS security profile based on:
+// https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility
+type ModernTLSProfile struct{}
+
+// CustomTLSProfile is a user-defined TLS security profile. Be extremely careful
+// using a custom TLS profile as invalid configurations can be catastrophic.
+type CustomTLSProfile struct {
+	TLSProfileSpec `json:",inline"`
+}
+
+// TLSProfileType defines a TLS security profile type.
+type TLSProfileType string
+
+const (
+	// Old is a TLS security profile based on:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Old_backward_compatibility
+	TLSProfileOldType TLSProfileType = "Old"
+	// Intermediate is a TLS security profile based on:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
+	TLSProfileIntermediateType TLSProfileType = "Intermediate"
+	// Modern is a TLS security profile based on:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility
+	TLSProfileModernType TLSProfileType = "Modern"
+	// Custom is a TLS security profile that allows for user-defined parameters.
+	TLSProfileCustomType TLSProfileType = "Custom"
+)
+
+// TLSProfileSpec is the desired behavior of a TLSSecurityProfile.
+type TLSProfileSpec struct {
+	// ciphers is used to specify the cipher algorithms that are negotiated
+	// during the TLS handshake.  For example, to use 3DES  (yaml):
+	//
+	//   ciphers:
+	//     - 3DES
+	//
+	Ciphers []string `json:"ciphers"`
```

Comment:
Are we doing no validation on this?  i.e. we expect the operators to all tolerate random junk here?  I am mostly trying to understand what happens if you manage to reduce the whitelist to something bad and then you cannot easily connect back to the kube API to undo the change.

Reply:
> Are we doing no validation on this? i.e. we expect the operators to all tolerate random junk here?

@enj will a validating admission webhook suffice?

Reply:
Sure but we will have to relax its validation over time which can have weird consequences on downgrade.  Also I do not think we can assume that all components can honor all ciphers.